File: apps/android/frick/src/main/java/dev/frick/client/FrickClientTelemetry.kt
package dev.frick.client

import kotlinx.serialization.json.JsonPrimitive

enum class FrickClientTelemetrySpanKind {
    CLIENT,
    INTERNAL,
}

enum class FrickClientTelemetrySpanStatus {
    OK,
    ERROR,
}

data class FrickClientTelemetrySpanStart(
    val name: String,
    val kind: FrickClientTelemetrySpanKind = FrickClientTelemetrySpanKind.CLIENT,
    val attributes: Map<String, JsonPrimitive> = emptyMap(),
)

data class FrickClientTelemetrySpanResult(
    val status: FrickClientTelemetrySpanStatus? = null,
    val attributes: Map<String, JsonPrimitive> = emptyMap(),
    val error: Throwable? = null,
)

interface FrickClientTelemetrySpan {
    val traceId: String?
        get() = null

    fun injectHeaders(headers: MutableMap<String, String>) = Unit

    fun setAttributes(attributes: Map<String, JsonPrimitive>) = Unit

    fun end(result: FrickClientTelemetrySpanResult? = null) = Unit
}

interface FrickClientTelemetryRuntime {
    fun startSpan(input: FrickClientTelemetrySpanStart): FrickClientTelemetrySpan

    fun recordCounter(name: String, value: Double, attributes: Map<String, JsonPrimitive> = emptyMap()) = Unit

    fun recordHistogram(name: String, value: Double, attributes: Map<String, JsonPrimitive> = emptyMap()) = Unit
}

object NoopFrickClientTelemetryRuntime : FrickClientTelemetryRuntime {
    override fun startSpan(input: FrickClientTelemetrySpanStart): FrickClientTelemetrySpan =
        NoopFrickClientTelemetrySpan
}

private object NoopFrickClientTelemetrySpan : FrickClientTelemetrySpan

internal fun startFrickClientTelemetrySpan(
    telemetry: FrickClientTelemetryRuntime,
    input: FrickClientTelemetrySpanStart,
): FrickClientTelemetrySpan =
    runCatching { telemetry.startSpan(input) }.getOrDefault(NoopFrickClientTelemetrySpan)

internal fun injectFrickClientTelemetryHeaders(
    span: FrickClientTelemetrySpan,
    headers: MutableMap<String, String>,
) {
    runCatching { span.injectHeaders(headers) }
}

internal fun getFrickClientTelemetryTraceId(span: FrickClientTelemetrySpan): String? =
    runCatching { span.traceId }.getOrNull()

internal fun finishFrickClientTelemetrySpan(
    span: FrickClientTelemetrySpan,
    result: FrickClientTelemetrySpanResult? = null,
) {
    runCatching { span.end(result) }
}

internal fun recordFrickClientTelemetryCounter(
    telemetry: FrickClientTelemetryRuntime,
    name: String,
    value: Double,
    attributes: Map<String, JsonPrimitive>,
) {
    runCatching { telemetry.recordCounter(name = name, value = value, attributes = attributes) }
}

internal fun recordFrickClientTelemetryHistogram(
    telemetry: FrickClientTelemetryRuntime,
    name: String,
    value: Double,
    attributes: Map<String, JsonPrimitive>,
) {
    runCatching { telemetry.recordHistogram(name = name, value = value, attributes = attributes) }
}

File: packages/swift/Sources/FrickSwift/FrickClientTelemetry.swift
import Foundation

public enum FrickClientTelemetryAttributeValue: Equatable, Sendable {
    case string(String)
    case int(Int)
    case double(Double)
    case bool(Bool)
}

public typealias FrickClientTelemetryAttributes = [String: FrickClientTelemetryAttributeValue]

public enum FrickClientTelemetrySpanKind: Equatable, Sendable {
    case client
    case internalOperation
}

public struct FrickClientTelemetrySpanStart: Equatable, Sendable {
    public let name: String
    public let kind: FrickClientTelemetrySpanKind
    public let attributes: FrickClientTelemetryAttributes

    public init(
        name: String,
        kind: FrickClientTelemetrySpanKind = .client,
        attributes: FrickClientTelemetryAttributes = [:]
    ) {
        self.name = name
        self.kind = kind
        self.attributes = attributes
    }
}

public enum FrickClientTelemetrySpanStatus: Equatable, Sendable {
    case ok
    case error
}

public struct FrickClientTelemetrySpanResult: Equatable, Sendable {
    public let status: FrickClientTelemetrySpanStatus?
    public let attributes: FrickClientTelemetryAttributes
    public let errorDescription: String?

    public init(
        status: FrickClientTelemetrySpanStatus? = nil,
        attributes: FrickClientTelemetryAttributes = [:],
        errorDescription: String? = nil
    ) {
        self.status = status
        self.attributes = attributes
        self.errorDescription = errorDescription
    }
}

public protocol FrickClientTelemetrySpan: Sendable {
    var traceId: String? { get }
    func injectHeaders(_ headers: inout [String: String]) throws
    func setAttributes(_ attributes: FrickClientTelemetryAttributes) throws
    func end(_ result: FrickClientTelemetrySpanResult?) throws
}

public extension FrickClientTelemetrySpan {
    var traceId: String? { nil }

    func injectHeaders(_ headers: inout [String: String]) throws {}

    func setAttributes(_ attributes: FrickClientTelemetryAttributes) throws {}

    func end(_ result: FrickClientTelemetrySpanResult?) throws {}
}

public protocol FrickClientTelemetryRuntime: Sendable {
    func startSpan(_ input: FrickClientTelemetrySpanStart) throws -> any FrickClientTelemetrySpan
    func recordCounter(name: String, value: Double, attributes: FrickClientTelemetryAttributes) throws
    func recordHistogram(name: String, value: Double, attributes: FrickClientTelemetryAttributes) throws
}

public extension FrickClientTelemetryRuntime {
    func recordCounter(name: String, value: Double, attributes: FrickClientTelemetryAttributes) throws {}

    func recordHistogram(name: String, value: Double, attributes: FrickClientTelemetryAttributes) throws {}
}

public struct FrickNoopClientTelemetryRuntime: FrickClientTelemetryRuntime {
    public init() {}

    public func startSpan(_ input: FrickClientTelemetrySpanStart) throws -> any FrickClientTelemetrySpan {
        FrickNoopClientTelemetrySpan()
    }
}

private struct FrickNoopClientTelemetrySpan: FrickClientTelemetrySpan {}

func startFrickClientTelemetrySpan(
    _ telemetry: any FrickClientTelemetryRuntime,
    _ input: FrickClientTelemetrySpanStart
) -> any FrickClientTelemetrySpan {
    do {
        return try telemetry.startSpan(input)
    } catch {
        return FrickNoopClientTelemetrySpan()
    }
}

func injectFrickClientTelemetryHeaders(
    _ span: any FrickClientTelemetrySpan,
    into headers: inout [String: String]
) {
    do {
        try span.injectHeaders(&headers)
    } catch {
        // Telemetry must never affect framework requests.
    }
}

func finishFrickClientTelemetrySpan(
    _ span: any FrickClientTelemetrySpan,
    _ result: FrickClientTelemetrySpanResult?
) {
    do {
        try span.end(result)
    } catch {
        // Telemetry must never affect framework behavior.
    }
}

func recordFrickClientTelemetryCounter(
    _ telemetry: any FrickClientTelemetryRuntime,
    name: String,
    value: Double,
    attributes: FrickClientTelemetryAttributes
) {
    do {
        try telemetry.recordCounter(name: name, value: value, attributes: attributes)
    } catch {
        // Telemetry must never affect framework behavior.
    }
}

func recordFrickClientTelemetryHistogram(
    _ telemetry: any FrickClientTelemetryRuntime,
    name: String,
    value: Double,
    attributes: FrickClientTelemetryAttributes
) {
    do {
        try telemetry.recordHistogram(name: name, value: value, attributes: attributes)
    } catch {
        // Telemetry must never affect framework behavior.
    }
}
